Npm package upload

sema-sli opened the pull request.

sema-sli committed:
Add npm legacy login feature

Users can now use the npm cli to log into an NPM repository via the
/pulp/api/v3/npm/cli/<repository> endpoint.

sema-sli committed:
Add npm publish feature

Users can now use the npm cli to push NPM packages to a repository via
the /pulp/api/v3/npm/cli/<repository> endpoint.

sema-sli committed:
Use settings.V3_API_ROOT to generate upload/repo hrefs in create@PackageViewSet

sema-mpi commented on pulp_npm/app/tasks/publishing.py:
```diff
@@ -0,0 +1,19 @@
+
+def publish(repository_pk, package_pk):
+    repository = NpmRepository.objects.get(pk=repository_pk)
+    package = Package.objects.get(pk=package_pk)
```
The `package` object should not be needed, in `new_version.add_content()` you only need the `package.pk` which should be just the passed `package_pk` to this function.

sema-mpi commented on pulp_npm/app/tasks/publishing.py:
```diff
@@ -0,0 +1,19 @@
+
+    # create new repository version
+    with repository.new_version(base_version=repository.latest_version()) as new_version:
+        new_version.add_content(Package.objects.filter(pk=package.pk))
```
See above, just use `package_pk` here.

sema-mpi commented on pulp_npm/app/tasks/publishing.py:
```diff
@@ -0,0 +1,19 @@
+    distribution = NpmDistribution.objects.get(repository=repository)
+
+    # create new repository version
+    with repository.new_version(base_version=repository.latest_version()) as new_version:
```
Why is `base_version=repository.latest_version()` needed here?

chrstphfrtz replied:
I have tested this in the conda plugin and as far as I can see it is not needed.

sema-mpi commented on pulp_npm/app/tasks/publishing.py:
```diff
@@ -0,0 +1,19 @@
+        new_version.add_content(Package.objects.filter(pk=package.pk))
+
+    # switch distribution to new version
+    distribution.repository_version = new_version
```
I'm not sure I fully understand how this works, but this also looks weird from a Python standpoint, isn't `new_version` here technically out of scope? I assume this is some Python internals and it "just works" but definitely raises an eyebrow for me.

sema-mpi commented on pulp_npm/app/utils.py:
```diff
@@ -1,6 +1,5 @@
 import re
 
-
```
This line removal should not be part of the commit.

sema-mpi commented on pulp_npm/app/viewsets.py:
```diff
@@ -43,60 +66,79 @@ class PackageViewSet(core.SingleArtifactContentUploadViewSet):
+        binary_data = base64.b64decode(attachment['data'])
+
+        # save upload
+        upload = Upload.objects.create(size=len(binary_data))
```
I would maybe double-check if we could utilize some buil-in Pulp class, etc. to not have to do this manually (the whole Upload/chunk thing), I would be surprised if there is not something already provided by Pulp.

chrstphfrtz replied:
You should be able to handle this without an upload object, right? It should be possible to use ```PulpTemporaryFile``` from ```pulpcore.plugin.models``` and then create the Artifact from that

```python
temp_file = PulpTemporaryFile(attachment['data'])
artifact = Artifact.from_pulp_temporary_file(temp_file)
```

sema-mpi commented on pulp_npm/app/viewsets.py:
```diff
@@ -43,60 +66,79 @@ class PackageViewSet(core.SingleArtifactContentUploadViewSet):
+        serializer = serializers.PackageSerializer(data=data)
+        # TODO fix when a better solution has been found
+        # change to pulp tempdir to work around incorrect path settings in the validator
+        os.chdir("/var/lib/pulp/tmp")
```
This really should not end up in production code, like really really not.
Especially since you never change the working directory back anywhere, so I guess for the lifetime of the process this is now the working directory (not sure how long this process here lives), so this seems super iffy.

chrstphfrtz replied:
Why is this actually needed? What fails when this is left out? Could the above mentioned ```PulpTemporaryFile``` solve this?

sema-mpi reviewed (commented):
I think as a first draft it is good and good job.
I don't have the full picture of how Pulp plugins should work/look and how objects and such relate to each other, but I see some things that struck me out as odd, especially the `os.chdir()` HACK.

sema-mpi commented on pulp_npm/app/tasks/publishing.py:
```diff
@@ -0,0 +1,19 @@
+        new_version.add_content(Package.objects.filter(pk=package.pk))
+
+    # switch distribution to new version
+    distribution.repository_version = new_version
```
When I remove the following lines completely:
```python
    distribution.repository_version = new_version
    distribution.save()
```
Then the `./tests/npm-repo.sh` script still passes.

However just adding a bit more logs and changing it pushing another package to the repository shows that the version increases from None -> 1 -> 2 after each respective new package push.

So I assume this might be needed indeed, but then the `npm-repo.sh` script should be updated to test the pushing and installing of two different packages.

(Also I'm fully aware that this is just me not understanding the internals of pulp)

chrstphfrtz replied:
I have also tested this in the conda plugin and it seems that the distribution always points to the latest repository anyways so this should not be needed as well.

sema-sli replied:
Yes I've tested with uploading two packages and it works. The version of the distribution updates automatically to the latest.

chrstphfrtz reviewed (commented):
I also think that this is a good first implementation of the requirements! I think most things have been said already but maybe check out the ```PulpTempFile``` instead of doing the upload manually. I don't know if this resolves your problems but I think its worth a try.

chrstphfrtz commented:
Maybe it is also a good idea to write a short documentation or workflow on how to use the plugin and on how to use the login, publish and pull-through cache features.

sema-sli committed:
Re-add accidentally removed newline in utils.py

sema-sli committed:
Remove unneeded package object from publishing task

sema-sli committed:
Remove base_version parameter in repository.new_version() in publish task

The new_version() method already takes the latest repository
version.

sema-sli committed:
Remove distribution version update in publish task

The update of the repository version in the distribution is done
automatically.

sema-sli committed:
Create package through artifact instead of upload in create@PackageViewset

sema-sli committed:
Remove repository from package validation data in create@PackageViewSet

The repository is not needed to validate the package data.
